Modules/Module_Processing/xASL_init_DefineStudyData.m

```diff
@@ -286,12 +302,18 @@
 % ------------------------------------------------------------------------------------------------
 %% 7) Parallelization: If running parallel, select cases for this worker
 if x.opts.nWorkers>1
-    nSubjPerWorker = ceil(x.dataset.nSubjects/x.opts.nWorkers); % ceil to make sure all subjects are processed
-    nSubjSessPerWorker = nSubjPerWorker*x.dataset.nSessions;
-    iStartSubject = (x.opts.iWorker-1)*nSubjPerWorker+1;
-    iEndSubject = min(x.opts.iWorker*nSubjPerWorker, x.dataset.nSubjects);
-    iStartSubjectSession = (x.opts.iWorker-1)*nSubjSessPerWorker+1;
-    iEndSubjectSession = min(x.opts.iWorker*nSubjSessPerWorker, x.dataset.nSubjectsSessions);
+    nSubjPerWorker = x.dataset.nSubjects/x.opts.nWorkers; % ceil to make sure all subjects are processed
+
+    % e.g., if nWorkers=3 & nSubjects=10
+    % iWorker 1 does [1 2 3]
+    % iWorker 2 does [4 5 6 7]
+    % iWorker 3 does [8 9 10]
+
+    iStartSubject = round((x.opts.iWorker-1)*nSubjPerWorker+1);
+    iEndSubject = min( round(x.opts.iWorker*nSubjPerWorker), x.dataset.nSubjects);
+
+    iStartSubjectSession = (iStartSubject-1) * x.dataset.nSessions + 1;

```

Comment:
I would prefer to remove this SubjectSession parameter completely. This is only going to be a source of errors for studies with uneven number of subjects!

We should always processdata based on subjects and for each subjects we go through sessions.

but the subjectsession thing cannot ever work. Best to remove it here to make sure it's not used anywhere. Especially when we are doing this cleaning.

Reply:
Also below - lets delete all mention of nSubjectSessions and SubjectSessions  everywhere.

Reply:
The SubjectSession parameter is generally used in a way that assumes we have the same number of sessions per subject - see line 315 for example. 

But this assumption won't hold anymore as we can have variable session numbers. So continuous counting of sessions across subjects won't match that formula anymore. Maybe this concept will still work and some SubjectSessions will be empty and skipped, or we might run into errors and mess up the processing order badly. We have previously seen some weird behavior and some scans not randomly processed - this could be the source of the error. 

In any case, we have argumented for Import and BIDS2legacy that we should split workers not for subjectsessions, but for subjects and then for each subject do all sessions. Even though it might happen that not all workers will then process the same amount of sessions.

So yes - I suggest to split workers to subjects. Then for each subject, we go across sessions/runs. And we definitively need to do this for BIDS2legacy and for import for various reasons. So to keep BIDS2legacy+processing in sync, we need to do this across subjects everywhere. This was supposed to be the major edit in this BIDS2legacy parallelization.

I'm of course happy to discuss this more.

Reply:
That's exactly the plan. 

At the start of the processing, `xASL_init_DefineStudyData` takes the maximum number of sessions and indeed uses this for all subjects. Missing sessions will be skipped, I fixed that in the past.
If you want to tweak our current use of x.nSessions and x.SESSIONS, such that x.SESSIONS varies between subjects, this will be a major overhaul of ExploreASL. Instead, what I suggest is that we keep this behavior and simply skip unexisting sessions.

The parallelization into different workers was already only split between subjects, and I will keep it this way.
So e.g.,

sub-01 has ASL_1 ASL_2 ASL_3
sub-02 has ASL_2 only

nWorkers is 2.

Then iWorker == 1 will do sub-01 including ASL_1 ASL_2 ASL_3
iWorker == 2 will do sub-02 structural, skip ASL_1, do ASL_2, skip ASL_3

Is this OK for you?

Reply:
OK. Makes sense. Having nSessions at max and skipping missing sessions is perfectly fine! The example described is then perfectly fine. OK for me. 

However, if it is like this, I do not understand what we need subjectSessions for. If we assume a unified number of sessions per subject and skipping of missing, then at each point of time, the conversion from subjectSession index to iSubject and iSession should be straightforward and the other way around. So the variable subjectSession has no added value and can only help us if added locally, but not globally.
Therefore, I would propose to get rid of it everywhere from the x-struct and at most keep at local function level or even remove from there.

Cause otherwise I'm afraid that it could be bug prone. Or do we need it for some other purpose?

Reply:
We use x.datasets.nSubjectsSessions 32 times in various parts of the code, and this is always x.dataset.nSubjects * x.dataset.nSessions. So it just saves a line of coding 32 times :)

We could change this but that could then actually induce bugs. At least I wouldn't do this in this issue :) OK?

Reply:
Indeed. A new issue #1537 created.